pkg/nodeagent/controller/healthcheck/kubelet.go

```diff
@@ -0,0 +1,271 @@
+			internalIP = addr.Address
+		default:
+			// ignore
+		}
+	}
+
+	if externalIP == "" && internalIP == "" {
+		if k.lastInternalIP.IsValid() {
+			k.recorder.Eventf(node, corev1.EventTypeWarning, "kubelet", "node status does neither have an internal nor an external IP, try to recover from last known internal IP:%s", k.lastInternalIP)
+		} else {
+			k.recorder.Event(node, corev1.EventTypeWarning, "kubelet", "node status does neither have an internal nor an external IP")
+		}
+		addresses, err := k.getAddresses()
+		if err != nil {
+			return fmt.Errorf("unable to list all network interface IP addresses")
+		}
+
+		for _, addr := range addresses {
+			parsed, err := netip.ParsePrefix(addr.String())
+			parsedIP := parsed.Addr()
+			if err != nil {
+				return fmt.Errorf("unable to parse IP address %w", err)
+			}
+			if parsedIP.Compare(k.lastInternalIP) != 0 {
+				continue
+			}
+
+			// One of the ip addresses on the node matches the previous set internalIP of the node which is now gone, set it again.
+			node.Status.Addresses = []corev1.NodeAddress{
+				{
+					Type:    corev1.NodeInternalIP,
+					Address: k.lastInternalIP.String(),
+				},
+			}
+
+			err = k.client.Status().Update(ctx, node)
+			if err != nil {
+				if !apierrors.IsConflict(err) {
+					log.Error(err, "Unable to update node with internal IP")
+					k.recorder.Eventf(node, corev1.EventTypeWarning, "kubelet", "unable to update node with internal IP: %s", err.Error())
+					return k.dbus.Restart(ctx, k.recorder, node, kubeletServiceName)
+				}
+				return err
+			}
+			log.Info("Updated internal IP address of node", "ip", k.lastInternalIP.String())
+			k.recorder.Eventf(node, corev1.EventTypeNormal, "kubelet", "updated the lost internal IP address of node to the previous known: %s ", k.lastInternalIP.String())
+		}
+	} else {
+		var err error
+		k.lastInternalIP, err = netip.ParseAddr(internalIP)
+		if err != nil {
+			return fmt.Errorf("unable to parse internal IP address %w", err)
+		}
+	}
+
+	return nil
+}
+
+// ToggleKubeletState should be triggered if the state of the kubelet changed from Ready -> NotReady or vise versa.
+// It returns true if a reboot of the node should be triggered.
+func (k *KubeletHealthChecker) ToggleKubeletState() bool {
+	k.KubeletReadynessToggles = append(k.KubeletReadynessToggles, time.Now())
+	if len(k.KubeletReadynessToggles) < maxToggles {
+		return false
+	}
+
+	// We have more than maxToggles reached but in a too long period.
+	if k.Clock.Since(k.KubeletReadynessToggles[0]).Abs() > toggleTimeSpan.Abs() {
+		k.KubeletReadynessToggles = k.KubeletReadynessToggles[:len(k.KubeletReadynessToggles)-1]
+		return false
+	}
+
+	// If the oldest entry is older than allowed, trim the slices.
+	if len(k.KubeletReadynessToggles) > maxToggles {
+		k.KubeletReadynessToggles = k.KubeletReadynessToggles[:len(k.KubeletReadynessToggles)-1]
+	}
+
+	// Reaching this point means we need to reboot.
+	clear(k.KubeletReadynessToggles)
```

Comment:
Why is the `if` block necessary at all when we `clear()` the slice anyway afterwards?